src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -710,6 +710,11 @@ async Task Core(MultiConnectSocketAsyncEventArgs internalArgs, Task<IPAddress[]>
                             if (address.AddressFamily == AddressFamily.InterNetworkV6)
                             {
                                 attemptSocket = tempSocketIPv6 ??= (Socket.OSSupportsIPv6 ? new Socket(AddressFamily.InterNetworkV6, socketType, protocolType) : null);
+                                if (attemptSocket is not null && address.IsIPv4MappedToIPv6)
+                                {
+                                    // We need a DualMode socket to connect to an IPv6-mapped IPv4 address.
+                                    attemptSocket.DualMode = true;
```

Comment:
If this address doesn't succeed, and this socket ends up getting reused for the next connection attempt with the next address, is this going to cause any problems?

Reply:
I'm not aware of any environment where this may cause troubles, and I believe `address.IsIPv4MappedToIPv6 == true` implies that `DnsEndpoint` is carrying an actual IPv6 string in most (if not 100%) of the cases, which means that there are no further addresses to test. To be extra-safe, we may consider remapping to IPv4 instead, but that would cost an extra allocation.

Note that this entire thing seems to be a rare corner case.

Reply:
Ok, thanks.  So a) we don't think it'll cause any problems and b) even if it did we don't think it can happen.

Reply:
> I believe address.IsIPv4MappedToIPv6 == true implies that DnsEndpoint is carrying an actual IPv6 string in most (if not 100%) of the cases

I agree, this should be true. Should we assert this?

Reply:
Aside: It seems like we could avoid some complexity in this code if we checked up front whether the DndEndpoint is actually just an IP address string. In particular, we can avoid all the retry logic since we know there's only one address to attempt to connect to, we can handle the special cases like this particular one, etc.

We end up doing this check as part of the name resolution anyway. Is it worth it to take that into consideration here?

Not something we need for this PR, of course -- but perhaps we should consider filing an issue to look at this in the future.

Reply:
Makes sense to me, not sure if the code would be simpler, but definitely better optimized. We can open a separate issue, if we decide to merge this PR.

Reply:
> Should we assert this?

I prefer to not do it in this PR (merging it as is), we can add well-thought assertions in a future refactor.